### ISSUE TITLE
sql: warn users away from updated at indexes

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#41058]
One thing we noticed when attempting to debug a recent transaction retry is that code ported over from Oracle can use “updated at indexes” to power efficient queries. In CockroachDB, these queries are messy as they can cause hotspots. This is true not just in CRDB but in all distributed systems. Updated at indexes cause multiple updates for the same range of data. Customers might like these in Oracle but we provide better alternatives that pair well with our underlying architecture. Oftentimes, we end up asking developers to remove these indexes from their application. We should provide a warning to users not use updated at indexes or perhaps even forbid them behind sql safe updates.

**Comment by solongordon** [comment]
Can you be more specific about what an "updated at" index is? Does this just mean an index on an `updated_at TIMESTAMPTZ` column?

**Comment by rkruze** [comment]
I've seen schemas like this in the past: 

```sql
CREATE TABLE t1 (id UUID PRIMARY KEY, updated_at TIMESTAMPTZ, INDEX updated_at_index (updated_at));
```

We should provide a warning or make it harder to create an index where the first part of the index is a timestamp as these always tend to be a hotspot or point of contention.

**Comment by rafiss** [comment]
cc @vy-ton -- should we just log a warning/notice if we ever see an index where the first column is a timestamp?

**Comment by michae2** [comment]
If the timestamp column has `DEFAULT current_timestamp()` it seems like even more of a red flag. (And I wonder if `DEFAULT unique_rowid()` is also a red flag, as this uses the current time internally.)

Maybe the user should be prompted to use a hash-sharded index for these cases?

**Comment by michae2** [comment]
> (And I wonder if `DEFAULT unique_rowid()` is also a red flag, as this uses the current time internally.)

Also see https://github.com/cockroachdb/cockroach/pull/54675 for `unordered_unique_rowid`.

**Comment by rafiss** [comment]
I agree to starting with adding this warning when a column has `DEFAULT current_timestamp()` and it appears as the first column in an index (either primary or secondary). I'll update the issue description.

**Comment by XiongKezhi** [comment]
Hi, I think this is a good start for me as a newcomer so I wonder if this is still an issue? Also, I'm not sure If I understand the context here: is it because of locality (the most recently updated ones are updated more often than others) that causes a hot spot?

**Comment by vy-ton** [comment]
fyi @devadvocado @nickvigilante this was the issue that overlaps with index recommendations for hash-sharded indexes

**Comment by michae2** [comment]
Adding SQL Queries team so that we can consider doing something with index recommendations.

**Comment by bodagovsky** [comment]
Hi everyone!
Just wanted to notify that I have started working on that issue

**Comment by bodagovsky** [comment]
What about `DATE` and `TIME` types? Do we want to shoot a notice on these as well?

**Comment by bodagovsky** [comment]
I will be ready to push a PR soon, so I would like to clarify:

- shall we aim for `TIMESTAMP` and `TIMESTAMPTZ` fields being indexed on the first place
- or there will be a notice only for those who has `DEFAULT current_timestamp()` clause?
- or both?
Why I am asking is because it is also possible to do something like this:
`updated_at int DEFAULT current_timestamp()::int`

**Comment by michae2** [comment]
@bodagovsky good point, I think it would make sense to check for these specific functions in the column's `DEFAULT` expression rather than checking the column type:

- `clock_timestamp`
- `current_date`
- `current_time`
- `current_timestamp`
- `localtimestamp`
- `now`
- `statement_timestamp`
- `transaction_timestamp`

**Comment by bodagovsky** [comment]
Sorry, if this one is messy, my first PR in open source
Hope I will manage to fix all your comments soon!

**Comment by bodagovsky** [comment]
cc @awoods187 @michae2
